### PR TITLE
Added AADFeatureRolloutPolicy resource

### PR DESCRIPTION
**Pull request by ritikmit**
#### Pull Request (PR) description
Adding new DSC Entra resource - AADFeatureRolloutPolicy

### CODE REVIEW

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+        [System.Boolean]
+        $IsAppliedToOrganization,
+
+        [Parameter()]
+        [System.Boolean]
+        $IsEnabled,
+
+        [Parameter(Mandatory = $true)]
+        [System.String]
+        $Id,
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1:31]: Parameter ID should not be mandatory

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+
+        [Parameter()]
+        [System.Boolean]
+        $IsAppliedToOrganization,
+
+        [Parameter()]
+        [System.Boolean]
+        $IsEnabled,
+
+        [Parameter(Mandatory = $true)]
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: Should not be mandatory

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+
+    $BoundParameters = Remove-M365DSCAuthenticationParameter -BoundParameters $PSBoundParameters
+
+
+    if ($Ensure -eq 'Present' -and $currentInstance.Ensure -eq 'Absent')
+    {
+        Write-Verbose -Message "Creating an Azure AD Policy Feature Rollout Policy with DisplayName {$DisplayName}"
+
+        $createParameters = ([Hashtable]$BoundParameters).Clone()
+        $createParameters = Rename-M365DSCCimInstanceParameter -Properties $createParameters
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: I don't think this is necessary since we don't seem to have any CIMInstance property defined.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+        $createParameters.Remove('Id') | Out-Null
+
+        $keys = (([Hashtable]$createParameters).Clone()).Keys
+        foreach ($key in $keys)
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: Same as above:
I don't think this is necessary since we don't seem to have any CIMInstance property defined.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+        #region resource generator code
+        $policy = New-MgBetaPolicyFeatureRolloutPolicy -BodyParameter $createParameters
+        #endregion
+    }
+    elseif ($Ensure -eq 'Present' -and $currentInstance.Ensure -eq 'Present')
+    {
+        Write-Verbose -Message "Updating the Azure AD Policy Feature Rollout Policy with Id {$($currentInstance.Id)}"
+
+        $updateParameters = ([Hashtable]$BoundParameters).Clone()
+        $updateParameters = Rename-M365DSCCimInstanceParameter -Properties $updateParameters
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: I don't think this is necessary since we don't seem to have any CIMInstance property defined.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+
+        $updateParameters.Remove('Id') | Out-Null
+        $updateParameters.Remove('Feature') | Out-Null
+
+        $keys = (([Hashtable]$updateParameters).Clone()).Keys
+        foreach ($key in $keys)
+        {
+            if ($null -ne $pdateParameters.$key -and $updateParameters.$key.GetType().Name -like '*CimInstance*')
+            {
+                $updateParameters.$key = Convert-M365DSCDRGComplexTypeToHashtable -ComplexObject $updateParameters.FeatureRolloutPolicyId
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: Is this necessary? I don't think the FeatureRolloutPolicyId property even exists on PSBoundParameters based on the expected parameters.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+
+        [Parameter()]
+        [System.Boolean]
+        $IsAppliedToOrganization,
+
+        [Parameter()]
+        [System.Boolean]
+        $IsEnabled,
+
+        [Parameter(Mandatory = $true)]
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: Should not be mandatory

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+    $testResult = $true
+
+    #Compare Cim instances
+    foreach ($key in $PSBoundParameters.Keys)
+    {
+        $source = $PSBoundParameters.$key
+        $target = $CurrentValues.$key
+        if ($null -ne $source -and $source.GetType().Name -like '*CimInstance*')
+        {
+            $testResult = Compare-M365DSCComplexObject `
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: I don't think these CIM Instances checks are necessary

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1 [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+            Write-Host "`r`n" -NoNewline
+        }
+        foreach ($config in $getValue)
+        {
+            $displayedKey = $config.Id
+            if (-not [String]::IsNullOrEmpty($config.displayName))
+            {
+                $displayedKey = $config.displayName
+            }
+            elseif (-not [string]::IsNullOrEmpty($config.name))
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.psm1]: $name is not a parameter of this object, therefore this will never work. Let's remove it and only keep DisplayName

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof [hunk 10]
```diff
@@ -0,0 +1,19 @@
+
+[ClassVersion("1.0.0.0"), FriendlyName("AADFeatureRolloutPolicy")]
+class MSFT_AADFeatureRolloutPolicy : OMI_BaseResource
+{
+    [Write, Description("A description for this feature rollout policy.")] String Description;
+    [Required, Description("The display name for this  feature rollout policy.")] String DisplayName;
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof]: This should actually be marked as a Key parameter instead of a Required one.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    [Write, Description("Possible values are: passthroughAuthentication, seamlessSso, passwordHashSync, emailAsAlternateId, unknownFutureValue, certificateBasedAuthentication. You must use the Prefer: include-unknown-enum-members request header to get the following value or values in this evolvable enum: certificateBasedAuthentication. For more information about the prerequisites for the enabled features, see Prerequisites for enabled features."), ValueMap{"passthroughAuthentication","seamlessSso","passwordHashSync","emailAsAlternateId","unknownFutureValue","certificateBasedAuthentication"}, Values{"passthroughAuthentication","seamlessSso","passwordHashSync","emailAsAlternateId","unknownFutureValue","certificateBasedAuthentication"}] String Feature;
+    [Write, Description("Indicates whether this feature rollout policy should be applied to the entire organization.")] Boolean IsAppliedToOrganization;
+    [Write, Description("Indicates whether the feature rollout is enabled.")] Boolean IsEnabled;
+    [Key, Description("The unique identifier for an entity. Read-only.")] String Id;
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof]: This should not be required and therefore should be a Write parameter.

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/settings.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+                                             ],
+                                    "update":  [
+
+                                               ]
+                                },
+                  "application":  {
+                                      "read":  [
+
+                                               ],
+                                      "update":  [
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/settings.json]: Required permissions are Directory.Read.All for Read or Directory.ReadWrite.All for Update

File: Modules/Microsoft365DSC/SchemaDefinition.json [hunk 13]
```diff
@@ -3519,6 +3519,81 @@
                            }
                        ]
     },
+    {
+        "ClassName":  "MSFT_AADFeatureRolloutPolicy",
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/SchemaDefinition.json:3523]: In the future, we don't need to touch this file. It gets generated automatically upon merging the PR.

File: ResourceGenerator/M365DSCResourceGenerator.psm1 [hunk 14]
```diff
@@ -1635,8 +1634,7 @@ function Get-Microsoft365DSCModuleCimClass
         $ResourceName
     )
 
-    Import-Module -Name Microsoft365DSC -Force
-    $modulePath = Split-Path -Path (Get-Module -Name Microsoft365DSC).Path
+    $modulePath = "C:\Program Files\WindowsPowerShell\Modules\Microsoft365DSC\Modules\Microsoft365DSC"
```
Comment (NikCharlebois) [inline ResourceGenerator/M365DSCResourceGenerator.psm1]: We should not touch this file here. Please undo the changes in the PR. The module can be installed anywhere on the dev's machine and we shouldn't hardcode paths.

File: Tests/Integration/Microsoft365DSC/M365DSCIntegration.AAD.Create.Tests.ps1 [hunk 15]
```diff
@@ -290,6 +290,19 @@
                     Principal       = "AdeleV@$TenantId";
                     RoleDefinition  = "Catalog creator";
                 }
+                AADFeatureRolloutPolicy 'AADFeatureRolloutPolicy-CertificateBasedAuthentication rollout policy'
```
Comment (NikCharlebois) [inline Tests/Integration/Microsoft365DSC/M365DSCIntegration.AAD.Create.Tests.ps1:293]: These tests are generated automatically based on the Example files and we shouldn't manually edit them.

**Review by NikCharlebois — CHANGES_REQUESTED** [review]
Several changes requested. Also, we need to add an Update and a Remove example, not just Create.

File: ResourceGenerator/M365DSCResourceGenerator.psm1 (hunk 14 above)
Comment (ritikmit) [reply ResourceGenerator/M365DSCResourceGenerator.psm1]: I thought I removed these changes, let me revert

File: Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof [hunk 16]
```diff
@@ -0,0 +1,19 @@
+
+[ClassVersion("1.0.0.0"), FriendlyName("AADFeatureRolloutPolicy")]
+class MSFT_AADFeatureRolloutPolicy : OMI_BaseResource
+{
+    [Write, Description("A description for this feature rollout policy.")] String Description;
+    [Key, Description("The display name for this  feature rollout policy.")] String DisplayName;
```
Comment (NikCharlebois) [inline Modules/Microsoft365DSC/DSCResources/MSFT_AADFeatureRolloutPolicy/MSFT_AADFeatureRolloutPolicy.schema.mof:6]: Looks like there is an extra space between this & feature

**Review by NikCharlebois — APPROVED** [review]
